qualtran/bloqs/mod_arithmetic/mod_multiplication_test.py

```diff
@@ -132,6 +135,64 @@ def test_examples_modmul(bloq_autotester):
     bloq_autotester(_modmul)
 
 
+def test_examples_dirtyoutofplacemontgomerymodmul_small(bloq_autotester):
+    bloq_autotester(_dirtyoutofplacemontgomerymodmul_small)
+
+
+def test_examples_dirtyoutofplacemontgomerymodmul_medium(bloq_autotester):
+    bloq_autotester(_dirtyoutofplacemontgomerymodmul_medium)
+
+
 @pytest.mark.notebook
 def test_notebook():
     qlt_testing.execute_notebook('mod_multiplication')
+
+
+@pytest.mark.parametrize('p', (7, 9, 11))
+@pytest.mark.parametrize('uncompute', [True, False])
+@pytest.mark.parametrize(
+    ['n', 'm'], [(n, m) for n in range(6, 10) for m in range(1, n + 1) if n % m == 0]
+)
+def test_dirtyoutofplacemontgomerymodmul_decomposition(n, m, p, uncompute):
+    b = DirtyOutOfPlaceMontgomeryModMul(n, m, p, uncompute)
+    qlt_testing.assert_valid_bloq_decomposition(b)
+
+
+@pytest.mark.parametrize('p', (7, 9, 11))
+@pytest.mark.parametrize('uncompute', [True, False])
+@pytest.mark.parametrize(
+    ['n', 'm'], [(n, m) for n in range(6, 10) for m in range(1, n + 1) if n % m == 0]
+)
+def test_dirtyoutofplacemontgomerymodmul_bloq_counts(n, m, p, uncompute):
+    b = DirtyOutOfPlaceMontgomeryModMul(n, m, p, uncompute)
+    qlt_testing.assert_equivalent_bloq_counts(b, [ignore_alloc_free, ignore_split_join])
+
+
+@pytest.mark.parametrize('uncompute', [True, False])
+def test_dirtyoutofplacemontgomerymodmul_symbolic_cost(uncompute):
+    n, m, p = sympy.symbols('n m p', integer=True)
+
+    # In Litinski 2023 https://arxiv.org/abs/2306.08585 a window size of 4 is used.
+    # The cost function generally has floor/ceil division that disappear for bitsize=0 mod 4.
+    # This is why instead of using bitsize=n directly, we use bitsize=4*m=n.
+    b = DirtyOutOfPlaceMontgomeryModMul(4 * m, 4, p, uncompute)
+    cost = get_cost_value(b, QECGatesCost()).total_t_and_ccz_count()
+    assert cost['n_t'] == 0
+
+    # Litinski 2023 https://arxiv.org/abs/2306.08585
+    # Figure/Table 8. Lists modular multiplication as 2.25n^2+9n toffoli.
+    # The following formula is 2.25n^2+8.25n-1 written with rationals because sympy comparison fails with floats.
+    assert isinstance(cost['n_ccz'], sympy.Expr)
+    assert (
+        cost['n_ccz'].subs(m, n / 4).expand()
+        == sympy.Rational(9, 4) * n**2 + sympy.Rational(33, 4) * n - 1
+    )
+
+
+@pytest.mark.parametrize('p', (3, 5, 7))
+@pytest.mark.parametrize(
+    ['n', 'm'], [(n, m) for n in range(5, 8) for m in range(1, n + 1) if n % m == 0]
+)
+def test_dirtyoutofplacemontgomerymodmul_classical_action(n, m, p):
+    b = DirtyOutOfPlaceMontgomeryModMul(n, m, p, False)
+    qlt_testing.assert_consistent_classical_action(b, x=range(1, p), y=range(1, p))
```

Comment:
I think we need a test for the SingleWindow one too, now

Reply:
done